### PR TITLE
System.Security: Remove last usages of DerSequenceReader/DerEncoder/AsnSerializer

### PR DESCRIPTION
**Pull request by filipnavara**
- Convert reminder of S.S.C.X509Certificates to use ASNXML/AsnReader/AsnWriter instead of DerSequenceReader/DerEncoder.
- Convert X509ExtensionAsn to code generator. Remove last usages of ASN serializer from code base.
- Remove DerEncoder, DerSequenceReader and related tests.

### CODE REVIEW

**Comment by filipnavara** [comment]
Has few small conflicts with #32161, will rebase and drop the WIP tag once that is solved and all the builds pass. This should be the last big PR in the series, so thanks for all the reviews so far :) I may follow up with some small clean-ups, but they should be much more localized and way smaller.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs [hunk 1]
```diff
@@ -210,6 +210,35 @@ public static void ValidateDer(ReadOnlyMemory<byte> encodedValue)
...
+                    // NULL character which was literally embedded in the DER would cause a
+                    // failure in .NET whereas it wouldn't have with strcmp.
+                    return tavReader.GetCharacterString((UniversalTagNumber)tag.TagValue).TrimEnd('\0');
+                    
+                default:
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+            }
+        }
+    }
+
```
Comment (filipnavara) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: The whole `ReadDirectoryString` method is just moved around from `X500NameEncoder`. The added `.TrimEnd('\0')` is needed to pass unit test, it was previously done inside `DerSequenceReader`. Technically `IA5String` is also not part of DictionaryString, but making that correction is outside the scope of the PR.

File: src/Common/src/System/Security/Cryptography/Asn1/X509ExtensionAsn.manual.cs [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.X509Certificates;
+
+namespace System.Security.Cryptography.Asn1
+{
+	internal partial struct X509ExtensionAsn
+    {
+        public X509ExtensionAsn(X509Extension extension)
+        {
```
Comment (filipnavara) [inline src/Common/src/System/Security/Cryptography/Asn1/X509ExtensionAsn.manual.cs]: The `skipCopy` parameter is removed and the copying of RawData is never done. The object is never used for anything but serialization with a lifetime within a single function.

File: src/System.Security.Cryptography.Pkcs/src/System.Security.Cryptography.Pkcs.csproj [hunk 3]
```diff
@@ -397,8 +394,16 @@
       <Link>Common\System\Security\Cryptography\Asn1\SubjectPublicKeyInfoAsn.xml.cs</Link>
       <DependentUpon>Common\System\Security\Cryptography\Asn1\SubjectPublicKeyInfoAsn.xml</DependentUpon>
     </Compile>
-    <Compile Include="$(CommonPath)\System\Security\Cryptography\Asn1\X509ExtensionAsn.cs">
-      <Link>Common\System\Security\Cryptography\Asn1\X509ExtensionAsn.cs</Link>
+    <AsnXml Include="$(CommonPath)\System\Security\Cryptography\Asn1\X509ExtensionAsn.xml">
+      <Link>Common\System\Security\Cryptography\Asn1\Rc2CbcParameters.xml</Link>
```
Comment (bartonjs) [inline src/System.Security.Cryptography.Pkcs/src/System.Security.Cryptography.Pkcs.csproj]: This `<Link>` value probably wants to say X509ExtensionAsn :smile:

Comment (bartonjs) [reply src/System.Security.Cryptography.Pkcs/src/System.Security.Cryptography.Pkcs.csproj]: (Similarly for the others)

Comment (filipnavara) [reply src/System.Security.Cryptography.Pkcs/src/System.Security.Cryptography.Pkcs.csproj]: Oh yeah, cut and paste error, will fix ;)

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedCertificateFinder.cs [hunk 4]
```diff
@@ -164,7 +135,8 @@ public void FindByTemplateName(string templateName)
                     if (ext != null)
                     {
                         // Try a V1 template structure, just a string:
-                        string decodedName = DerStringToManagedString(ext.RawData);
+                        AsnReader reader = new AsnReader(ext.RawData, AsnEncodingRules.DER);
+                        string decodedName = reader.ReadDirectoryString();
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/ManagedCertificateFinder.cs]: ThrowIfNotEmpty?

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificatePolicy.cs [hunk 5]
```diff
@@ -86,10 +83,8 @@ internal bool MatchesCertificatePolicies(Oid policyOid)
 
                 if (policy.PolicyMapping != null)
                 {
-                    for (int iMapping = 0; iMapping < policy.PolicyMapping.Count; iMapping++)
+                    foreach (CertificatePolicyMappingAsn mapping in policy.PolicyMapping)
                     {
-                        CertificatePolicyMapping mapping = policy.PolicyMapping[iMapping];
-
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificatePolicy.cs]: I guess no particular reason except that `iMapping`  was not used anywhere, so `foreach` seemed to show the intention better. I can revert this change.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs [hunk 6]
```diff
@@ -356,52 +326,24 @@ private static string FindAltNameMatch(byte[] extensionBytes, GeneralNameType ma
 
         private static IEnumerable<KeyValuePair<string, string>> ReadReverseRdns(X500DistinguishedName name)
         {
-            DerSequenceReader x500NameReader = new DerSequenceReader(name.RawData);
-            var rdnReaders = new Stack<DerSequenceReader>();
+            AsnReader x500NameReader = new AsnReader(name.RawData, AsnEncodingRules.DER);
+            AsnReader sequenceReader = x500NameReader.ReadSequence();
+            var rdnReaders = new Stack<AsnReader>();
```
[the remark this answers is not in the document]
Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs]: It is, I will fix it.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/CertificateTemplateAsn.xml [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+
+    CertificateTemplate ::= SEQUENCE {
+        templateID OBJECT IDENTIFIER,
+        templateMajorVersion TemplateVersion,
+        templateMinorVersion TemplateVersion OPTIONAL
+    }
+  -->
+  <asn:ObjectIdentifier name="TemplateID" backingType="string" />
+  <asn:Integer name="TemplateMajorVersion" backingType="int" />
+  <asn:Integer name="templateMinorVersion" backingType="int" optional="true" />
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/CertificateTemplateAsn.xml]: `templateMinorVersion` => `TemplateMinorVersion`

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/DistributionPointAsn.xml [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+  <!--
+    https://tools.ietf.org/html/rfc5280#section-4.2.1.13
+
+    DistributionPoint ::= SEQUENCE {
+        distributionPoint       [0]     DistributionPointName OPTIONAL,
+        reasons                 [1]     ReasonFlags OPTIONAL,
+        cRLIssuer               [2]     GeneralNames OPTIONAL
+    }
+  -->
+  <asn:AsnType name="DistributionPoint" typeName="System.Security.Cryptography.X509Certificates.Asn1.DistributionPointNameAsn" explicitTag="0" optional="true" />
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/DistributionPointAsn.xml]: DistributionPoint is defined in ASN.1 module section A.2, which is `DEFINITIONS IMPLICIT TAGS`, so it seems like this one should be implicitTag, not explicitTag (like the other two)

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificatePolicy.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/CertificatePolicy.cs]: The for loop is both stack smaller (by one int and maybe one T) and more performant (by one &&, one field read, and one ==).  Certainly that level of perf doesn't really matter here (given that policy mappings are... "rare" would imply more of them than I've seen)... but since the change isn't necessary (and doesn't have a perf reason), please undo it.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/ReasonFlagsAsn.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Runtime.InteropServices;
+using System.Security.Cryptography.Asn1;
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/ReasonFlagsAsn.cs]: Nit: It doesn't seem like either of these using-import statements is actually used.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs (hunk 1 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: `ReadDirectoryOrIA5String`?.  `ReadTrimmedDirectoryOrIA5String`?

The TrimEnd certainly matches what I did in DerSequenceReader... though it clearly should have only read up to the first '\0', not just trimmed ones off the end.  I'm sort of on the fence of whether we should leave it as is or fix it to match the comment.  But probably as a followup, since it bothers me to have functional changes and refactorings in the same checkin.  :smile:

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/DistributionPointAsn.xml (hunk 8 above)
Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/DistributionPointAsn.xml]: `DistributionPointNameAsn` is a choice and choices are always explicit.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/ReasonFlagsAsn.cs (hunk 9 above)
Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/ReasonFlagsAsn.cs]: You are right, I forgot to remove it.

File: src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/DistributionPointAsn.xml (hunk 8 above)
Comment (bartonjs) [reply src/System.Security.Cryptography.X509Certificates/src/System/Security/Cryptography/X509Certificates/Asn1/DistributionPointAsn.xml]: Ah, yes, that one :smile:.  Carry on.

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs (hunk 1 above)
Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Helpers.cs]: I'd prefer to make a follow up PR to clean this up. I have some ideas on how to make it cleaner :-)

File: src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs [hunk 10]
```diff
@@ -356,52 +326,25 @@ private static string FindAltNameMatch(byte[] extensionBytes, GeneralNameType ma
...
-                    }
-
-                    if (value != null)
-                    {
-                        yield return new KeyValuePair<string, string>(oid, value);
-                    }
+                    AsnReader tavReader = rdnReader.ReadSequence();
+                    string oid = tavReader.ReadObjectIdentifierAsString();
+                    string value = tavReader.ReadDirectoryOrIA5String();
+                    yield return new KeyValuePair<string, string>(oid, value);
```
Comment (bartonjs) [inline src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs]: Should there be a `tavReader.ThrowIfNotEmpty();` here?

Comment (filipnavara) [reply src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/CertificateData.cs]: Yep

**Comment by bartonjs** [comment]
@dotnet-bot Test NETFX x86 Release Build please
Test Packaging All Configurations x64 Debug Build please
Test Windows x64 Debug Build please

All failed with "Remote call on JNLP4-connect connection ... failed. The channel is closing down or has closed down"

**Comment by bartonjs** [comment]
Thanks, @filipnavara.

Farewell, DerSequenceReader.

**Comment by filipnavara** [comment]
Thanks for careful reviews, @bartonjs! Let's productise AsnReader/AsnWriter next... :-)
